src/openms/source/ANALYSIS/ID/FragmentIndex.cpp

```diff
@@ -252,6 +270,156 @@ namespace OpenMS
+          // the following part is 1:1 from sage
+          size_t idx = hit.peptide_idx - candidates_range.first;
+
+          auto& source = candidates.hits_[idx];
+          if (source.num_matched_ == 0)
+          {
+            ++candidates.scored_candidates_;
+            source.precursor_charge_ = precursor_charge;
+            source.peptide_idx_ = hit.peptide_idx;
+            source.isotope_error_ = isotope_error;
+          }
+          ++source.num_matched_;
+          ++candidates.matched_peaks_;
+        }
+      }
+  }
+
+  void FragmentIndex::trimHits(OpenMS::FragmentIndex::SpectrumMatchesTopN& init_hits) const
+  {
+      if (max_processed_hits_ < init_hits.hits_.size())
+      {
+        std::partial_sort(init_hits.hits_.begin(), init_hits.hits_.begin() + max_processed_hits_, init_hits.hits_.end(), [](SpectrumMatch a, SpectrumMatch b){
+          return a.num_matched_ > b.num_matched_;
+        });
+
+        init_hits.hits_.resize(max_processed_hits_);
+      }
+      for(auto hit_iter = init_hits.hits_.begin(); hit_iter != init_hits.hits_.end();){
+        if(hit_iter->num_matched_ < min_matched_peaks_)
+        {
+          init_hits.hits_.erase(hit_iter);
+        }else
+        {
+          ++hit_iter;
+        }
+      }
+  }
+
+  void FragmentIndex::openSearch(const OpenMS::MSSpectrum& spectrum, float precursor_mass, OpenMS::FragmentIndex::SpectrumMatchesTopN& sms, uint16_t charge)
+  {
+      pair<float, float> precursor_window;
+      if (open_precursor_window == 0)
+        precursor_window = make_pair(-precursor_mass * 0.05, precursor_mass * 0.01);
+      else
+        precursor_window = make_pair(-open_precursor_window, open_precursor_window); // sage took (-500, 100) fixed
+      OPENMS_LOG_WARN << "The Precursor mass window was set to: " << precursor_window.first << " " << precursor_window.second << endl;
+      auto candidates_range = getPeptidesInPrecursorRange(precursor_mass, precursor_window);
+
+      // adjust the window. with each decreasing peak the prob. that we have all modifications inside gets lower
+      // the linear decrease of the window below is a very coarse approximation and must be improved
+      pair<float, float> fragment_window;
+
+      SpectrumMatchesTopN hits_per_window;
+      hits_per_window.hits_.resize(candidates_range.second - candidates_range.first + 1);
+
+      for (auto peak = spectrum.end() - 1; peak >= spectrum.begin(); peak--)
+      {
+        queryPeak(hits_per_window, *peak, candidates_range, 0, charge);
+      }
+      trimHits(hits_per_window);
+      sms += hits_per_window;
+  }
+
+  void FragmentIndex::closedSearch(const OpenMS::MSSpectrum& spectrum, float mz, OpenMS::FragmentIndex::SpectrumMatchesTopN& sms, uint16_t charge)
+  {
+      for (int16_t isotope_error = min_isotope_error_; isotope_error <= max_isotope_error_; isotope_error++)
+      {
+        SpectrumMatchesTopN candidates_iso_error;
+        mz += (float)isotope_error *  (float)Constants::NEUTRON_MASS;
```

Comment:
looks wrong.
You are adding 0, 1, 2 etc to the m/z every time, but m/z is never reset.
Charge is neglected? Or is this uncharged mass? In which case `mz` is not the best name.
Also NEUTRON_MASS (you meant NEUTRON_MASS_U???) is probably better replaced with C13C12_MASSDIFF_U since isotope m/z's are dominated by carbon.